File: pine-java/src/main/java/page/liam/pine/operators/ObserveLog.java
package page.liam.pine.operators;

import com.fasterxml.jackson.databind.ObjectMapper;
import page.liam.pine.*;

import java.util.*;

/**
 * Operator: observe_log
 * Metadata contract
 *   CommonInput:  [<fields to observe>]
 *   CommonOutput: []
 *   ItemInput:    [<fields to observe>]
 *   ItemOutput:   []
 */
public class ObserveLog extends AbstractOperator {
    private static final ObjectMapper mapper = new ObjectMapper();
    private String prefix = "";

    @Override
    public void init(OperatorParams params) {
        Object v = params.get("log_prefix");
        if (v instanceof String) {
            prefix = (String) v;
        }
    }

    @Override
    public void execute(CancellationToken token, OperatorInput input, OperatorOutput output) {
        Map<String, Object> snapshot = new TreeMap<>();

        if (!commonInput().isEmpty()) {
            Map<String, Object> common = new TreeMap<>();
            for (String k : commonInput()) {
                common.put(k, input.common(k));
            }
            snapshot.put("common", common);
        }

        if (!itemInput().isEmpty() && input.itemCount() > 0) {
            int n = input.itemCount();
            List<Map<String, Object>> items = new ArrayList<>(n);
            for (int i = 0; i < n; i++) {
                items.add(new TreeMap<>());
            }
            for (String k : itemInput()) {
                Object[] col = input.itemColumn(k);
                for (int i = 0; i < n; i++) {
                    items.get(i).put(k, col[i]);
                }
            }
            snapshot.put("items", items);
        }

        try {
            String data = mapper.writeValueAsString(snapshot);
            if (!prefix.isEmpty()) {
                logf("[observe_log] %s %s", prefix, data);
            } else {
                logf("[observe_log] %s", data);
            }
        } catch (Exception e) {
            logf("[observe_log] %s marshal error: %s", prefix, e.getMessage());
        }
    }
}

File: pine-python/pine/operators/observe_log.py
from __future__ import annotations

import json
import sys

from pine.cancellation import CancellationToken
from pine.operator import (
    AbstractOperator,
    ConsumesRowSet,
    OperatorInput,
    OperatorOutput,
    OperatorParams,
)


class ObserveLog(AbstractOperator, ConsumesRowSet):
    def __init__(self):
        self._prefix = ""

    def init(self, params: OperatorParams):
        v = params.get("log_prefix")
        if isinstance(v, str):
            self._prefix = v

    def execute(
        self, token: CancellationToken, input_: OperatorInput,
        output: OperatorOutput,
    ) -> None:
        snapshot: dict = {}

        if self.common_input():
            common: dict = {}
            for k in self.common_input():
                common[k] = input_.common(k)
            snapshot["common"] = common

        if self.item_input() and input_.item_count() > 0:
            items: list = []
            for i in range(input_.item_count()):
                row: dict = {}
                for k in self.item_input():
                    row[k] = input_.item(i, k)
                items.append(row)
            snapshot["items"] = items

        try:
            data = json.dumps(snapshot, ensure_ascii=False, default=str)
            if self._prefix:
                print(f"[observe_log] {self._prefix} {data}", file=sys.stderr)
            else:
                print(f"[observe_log] {data}", file=sys.stderr)
        except (TypeError, ValueError, OverflowError) as e:
            print(f"[observe_log] {self._prefix} marshal error: {e}", file=sys.stderr)
